Fixed unique username issue #8

brw32 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

Added "unique=True" to username declaration in models.py

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Enforce uniqueness instead of manually checking.
<!--- If it fixes an open issue, please link to the issue here. -->https://github.com/docsocsf/mad/issues/8

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
Ran test.sh
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Set up virtualenv per instructions.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **README** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Commit:
Enforce unique user names in the database model

Set unique=TRUE and deleted TODO comment line

martinzlocha reviewed (changes requested):
Please change the commit message. Messages should be of the form: "This commit will **X**" where **X** is the actual message.

Eg. This commit will **Update the testing section in README**

To change a commit **last commit** which was pushed to remote use:
```
git commit --amend
git push --force <repository> <branch>
```

Also remove the ToDo message as it is unnecessary. All commits should be squashed into one. 

In the future don't copy the URL of an issue but reference it using a ```#<number of issue>``` eg. #8 and when ticking check-boxes make sure you don't leave spaces inside brackets as it will break the formatting.

Otherwise the fix is correct.

brw32 commented:
Hi Martin,
I changed the commit message and removed the ToDo message.
Thanks for your other comments -- I will be more careful in the future.

As you can probably tell, I am new to the github environment, and really 
appreciate your input.

Thanks, Brian

martinzlocha commented:
Hi Bryan,

I don't see any changes, have to pushed your changes successfully to the correct repository? You will have to [force push](https://stackoverflow.com/questions/5509543/how-do-i-properly-force-a-git-push) for the changes to be applied given that you are overwriting and existing commit.

![screenshot](https://user-images.githubusercontent.com/6172074/27509030-31cfba9e-58eb-11e7-8e98-a96a9b0a3b23.png)

We are happy to help if you need any advice. All repositories under DoCSoc SF are aimed at educating people.

Martin

brw32 commented:
Sorry, tried
1. "git push --force https://github.com/docsocsf/mad.git unique_usernames"
Got error: remote: Permission to docsocsf/mad.git denied to bwittig32.
2. Then tried "git push origin unique_usernames -f"
Got: Everything up to date.

Not sure what to do next. Another pull request?

martinzlocha commented:
It is bad practice to submit more pull requests. 

- First you have to commit the new changes that remove the ToDo.
- Then you git rebase -i HEAD~2 (this will rebase the last 2 commits). I would suggest you read up on [rebasing](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) first. This will allow you to **squash** the old and the new commit into one with a new message which fits the style mentioned above. 
- Afterwards you can push --force into origin unique_usernames.

brw32 commented:
I did the rebase.  Worked fine.  Then the push.

I didn't commit the deletion of the TODO line originally -- thought the commit --amend would do that, but realize it only referred to the current commit.

I am learning - thanks to you.

martinzlocha commented:
You still have an issue that there are 2 commits with bad commit messages. I think you haven't rebased successfully.

When you do ```rebase -i HEAD~2``` you should see an editor window with two commits both will have **pick** written next to them. Instead you should change the first one to **reword** so that you can change the commit message and the second one to **squash** so that the two commit messages get squashed into one.

After this you will be able to ```push --force```

Make sure to check the commit messages have updated [here](https://github.com/bwittig32/mad/commits/unique_usernames)

brw32 commented:
I did the rebase with the "reword" and "squash"; let's see if this works.  Thanks.

martinzlocha commented:
You have squashed them successfully so now its only one commit which is great but you have not changed the commit message. Instead you have merged the commit messages from the two commits:

![screenshot](https://user-images.githubusercontent.com/6172074/27512930-847a908e-594a-11e7-8e42-bc8e2a2bac64.png)

To change the last commit message again run ```git commit --amend``` which will open an editor window where you will see the message from the picture above. Change the message so that the first line is more meaningful eg. "Enforce unique usernames in the database model" leaving second line empty and then a one short  paragraph with the description (not two paragraphs like now).

Then you will ```push --force``` once again to force this change.

brw32 commented:
Amended the comment, as per your instructions (thanks!).  Hope I finally got it right.

martinzlocha commented:
Umm its back to 2 commits don't know why. Are you making the changes from different computers?

![screenshot](https://user-images.githubusercontent.com/6172074/27516038-a4f0de74-59b1-11e7-874f-2c48baf9c23a.png)

You will  have to follow the above instructions to ```rebase -i HEAD~2``` and correctly rename the commits. To make sure you have done it correctly you can use ```git log -2``` to see the last 2 commits. If its correct you should see one from you and one from me.

brw32 commented:
I did the "git log -2" and got results below.  I rebased and got the 2 "picks" below, but wasn't sure if I should change "pick" to something else or not. I left it alone and rebase finished ok. and tried to push but got Everything Up to Date.  

I am only using one computer.  Sorry there are so many pilot errors , but hard to tell what git is doing
behind the scenes.

---------------------------------------------------------------------------

PS C:\Users\Brian\documents\github\mad> git log -2
commit aeafebbb2bb5ddf4e2d2ddd47cd16d8ed515ac1b (HEAD)
Author: Brian R. Wittig <brian@brianwittig.com>
Date:   Fri Jun 23 16:55:48 2017 -0400

    Enforce unique user names in the database model

    Set unique=TRUE and deleted TODO comment line

commit 258c894f97267ee30977fd4868dc18f233019203 (origin/master, origin/HEAD, master)
Author: Martin Zlocha <mz4315@ic.ac.uk>
Date:   Thu Jun 22 22:06:41 2017 +0100

    Update the testing section in README

    README now encourages users to set up Circle CI on their forked repository to ensure that all tests pass.

martinzlocha commented:
Try running ```git push origin unique_usernames --force```

That should force the commits onto unique_usernames and rewrite the commits which are there.

brw32 commented:
Tried: git push origin unique_usernames --force.  Result: Everything up to date.
Instead ran: git push origin HEAD:unique_usernames --force.  That seemed to work.

We'll see.

brw32 commented:
Result from previous comment:

------------------------------------------------------------

PS C:\Users\Brian\Documents\github\mad> git push origin HEAD:unique_usernames --force
Counting objects: 4, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (4/4), done.
Writing objects: 100% (4/4), 437 bytes | 0 bytes/s, done.
Total 4 (delta 3), reused 0 (delta 0)
remote: Resolving deltas: 100% (3/3), completed with 3 local objects.
To https://github.com/bwittig32/mad.git
 + 940d208...aeafebb HEAD -> unique_usernames (forced update)

martinzlocha commented:
Very good, everything is as it should be 👍

If you have any more questions please ask.

brw32 commented:
Hallelujah!  I learned a lot -- thanks so much, Martin!
